### DISCUSSION TITLE
select daughterboard on usrp via soapysdr

### DISCUSSION THREAD

**Discussion by PripyatAutomations** [AlexandreRouma/SDRPlusPlus#1075] [Q&A] [answered]
in most soapysdr applications, you can pass a device string to it. Is there a way to pass the subdev=B0 to my usrp x300? Be nice to access HF in sdr++. Great program, thanks!

**Comment by AlexandreRouma** [comment, marked as answer]
Hi,

The SoapySDR source doesn't allow for these types of settings and support is not planned. However:

I've been working with Ettus Research to support all USRPs natively. Due to issues on some distros, the module is not currently included in prebuilt versions, but you can build SDR++ from source and enable the `usrp_source` module (add `-DOPT_BUILD_USRP_SOURCE=ON` to the cmake arguments). It's currently fully tested and working with the x300 series (I test with an x310 here). All combinations of daughterboards are supported.

**Comment by PripyatAutomations** [comment]
Thanks so much! I'll try it tonight.

**Reply by PripyatAutomations** [reply]
Rebuilt and i see during startup [22/05/2023 21:59:28.000] [INFO] Loading /usr/lib/sdrpp/plugins/usrp_source.so but no messages for post-init, etc nor entry in Sources list for uhd/usrp. tried removing ~/.config/sdrpp.

**Reply by AlexandreRouma** [reply]
that is normal. Since it's not included in official builds it's not instantiated by default either. Go in the module manager menu and create an instance of the `usrp_source` module.

**Reply by PripyatAutomations** [reply]
Got that done, but as soon as selecting the new source i get a crash - Let me know if i can do anything to aid in debugging. Appreciate your help Alexandre!

[22/05/2023 22:54:58.000] [INFO] AudioSourceModule 'Audio Source': Menu Deselect!
terminate called after throwing an instance of 'std::runtime_error'
  what():  Key already exists
Aborted

bt
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at ./nptl/pthread_kill.c:44
#1  0x00007ffff76a9d2f in __pthread_kill_internal (signo=6, threadid=<optimized out>) at ./nptl/pthread_kill.c:78
#2  0x00007ffff765aef2 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007ffff7645472 in __GI_abort () at ./stdlib/abort.c:79
#4  0x00007ffff6a9d919 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff6aa8e1a in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007ffff6aa8e85 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#7  0x00007ffff6aa90d8 in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#8  0x00007fffeceb4436 in OptionList<int, double>::define(int const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, double const&) () from /usr/lib/sdrpp/plugins/usrp_source.so
#9  0x00007fffeceb7682 in USRPSourceModule::selectChannel(std::shared_ptr<uhd::usrp::multi_usrp>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) () from /usr/lib/sdrpp/plugins/usrp_source.so
#10 0x00007fffeceb7bb6 in USRPSourceModule::selectChannel(std::shared_ptr<uhd::usrp::multi_usrp>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) () from /usr/lib/sdrpp/plugins/usrp_source.so
#11 0x00007fffeceb8d3b in USRPSourceModule::select(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
    () from /usr/lib/sdrpp/plugins/usrp_source.so
#12 0x00007fffeceb89f0 in USRPSourceModule::select(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
    () from /usr/lib/sdrpp/plugins/usrp_source.so
#13 0x00007fffecec04e2 in USRPSourceModule::menuSelected(void*) () from /usr/lib/sdrpp/plugins/usrp_source.so
#14 0x00007ffff79cc1a4 in SourceManager::selectSource(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) () from /home/radio/SDRPlusPlus/build/core/libsdrpp_core.so
#15 0x00007ffff78ccf19 in sourcemenu::selectSource(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
    () from /home/radio/SDRPlusPlus/build/core/libsdrpp_core.so
#16 0x00007ffff78cdcad in sourcemenu::draw(void*) () from /home/radio/SDRPlusPlus/build/core/libsdrpp_core.so
#17 0x00007ffff7904b80 in Menu::draw(bool) () from /home/radio/SDRPlusPlus/build/core/libsdrpp_core.so

**Reply by AlexandreRouma** [reply]
can you run `uhd_usrp_probe` and send back everything it prints?

**Reply by PripyatAutomations** [reply]
/
|       Device: X-Series Device
|     _____________________________________________________
|    /
|   |       Mboard: X300
|   |   revision: 4
|   |   product: 30518
|   |   mac-addr0: 00:80:2f:0a:ea:60
|   |   mac-addr1: 00:80:2f:0a:ea:61
|   |   gateway: 192.168.10.1
|   |   ip-addr0: 10.11.0.11
|   |   subnet0: 255.255.255.0
|   |   ip-addr1: 192.168.20.2
|   |   subnet1: 255.255.255.0
|   |   ip-addr2: 192.168.30.2
|   |   subnet2: 255.255.255.0
|   |   ip-addr3: 192.168.40.2
|   |   subnet3: 255.255.255.0
|   |   serial: F4D29E
|   |   FW Version: 6.1
|   |   FPGA Version: 39.1
|   |   FPGA git hash: 92c09f7
|   |   RFNoC capable: Yes
|   |   
|   |   Time sources:  internal, external, gpsdo
|   |   Clock sources: internal, external, gpsdo
|   |   Sensors: gps_servo, gps_time, gps_locked, gps_gprmc, gps_gpgga, ref_locked
|     _____________________________________________________
|    /
|   |       RFNoC blocks on this device:
|   |   
|   |   * 0/DDC#0
|   |   * 0/DDC#1
|   |   * 0/DUC#0
|   |   * 0/DUC#1
|   |   * 0/Radio#0
|   |   * 0/Radio#1
|   |   * 0/Replay#0
|     _____________________________________________________
|    /
|   |       Static connections on this device:
|   |   
|   |   * 0/SEP#0:0==>0/DUC#0:0
|   |   * 0/DUC#0:0==>0/Radio#0:0
|   |   * 0/Radio#0:0==>0/DDC#0:0
|   |   * 0/DDC#0:0==>0/SEP#0:0
|   |   * 0/Radio#0:1==>0/DDC#0:1
|   |   * 0/DDC#0:1==>0/SEP#1:0
|   |   * 0/SEP#2:0==>0/DUC#1:0
|   |   * 0/DUC#1:0==>0/Radio#1:0
|   |   * 0/Radio#1:0==>0/DDC#1:0
|   |   * 0/DDC#1:0==>0/SEP#2:0
|   |   * 0/Radio#1:1==>0/DDC#1:1
|   |   * 0/DDC#1:1==>0/SEP#3:0
|   |   * 0/SEP#4:0==>0/Replay#0:0
|   |   * 0/Replay#0:0==>0/SEP#4:0
|   |   * 0/SEP#5:0==>0/Replay#0:1
|   |   * 0/Replay#0:1==>0/SEP#5:0
|     _____________________________________________________
|    /
|   |       TX Dboard: 0/Radio#0
|   |   ID: WBX (0x0052)
|   |     _____________________________________________________
|   |    /
|   |   |       TX Frontend: 0
|   |   |   Name: WBXv2 TX+GDB
|   |   |   Antennas: TX/RX, CAL
|   |   |   Sensors: lo_locked
|   |   |   Freq range: 68.750 to 2200.000 MHz
|   |   |   Gain range PGA0: 0.0 to 25.0 step 0.1 dB
|   |   |   Bandwidth range: 40000000.0 to 40000000.0 step 0.0 Hz
|   |   |   Connection Type: IQ
|   |   |   Uses LO offset: No
|     _____________________________________________________
|    /
|   |       RX Dboard: 0/Radio#0
|   |   ID: WBX, WBX + Simple GDB (0x0053)
|   |     _____________________________________________________
|   |    /
|   |   |       RX Frontend: 0
|   |   |   Name: WBXv2 RX+GDB
|   |   |   Antennas: TX/RX, RX2, CAL
|   |   |   Sensors: lo_locked
|   |   |   Freq range: 68.750 to 2200.000 MHz
|   |   |   Gain range PGA0: 0.0 to 31.5 step 0.5 dB
|   |   |   Bandwidth range: 40000000.0 to 40000000.0 step 0.0 Hz
|   |   |   Connection Type: IQ
|   |   |   Uses LO offset: No
|     _____________________________________________________
|    /
|   |       TX Dboard: 0/Radio#1
|   |   ID: LF TX (0x630e)
|   |   Serial: 3071CB4
|   |     _____________________________________________________
|   |    /
|   |   |       TX Frontend: 0
|   |   |   Name: LFTX (0)
|   |   |   Antennas: AB, BA, A, B
|   |   |   Sensors: 
|   |   |   Freq range: -32.000 to 32.000 MHz
|   |   |   Gain Elements: None
|   |   |   Bandwidth range: 64000000.0 to 64000000.0 step 0.0 Hz
|   |   |   Connection Type: IQ
|   |   |   Uses LO offset: No
|     _____________________________________________________
|    /
|   |       RX Dboard: 0/Radio#1
|   |   ID: LF RX (0x630f)
|   |   Serial: F5770A
|   |     _____________________________________________________
|   |    /
|   |   |       RX Frontend: 0
|   |   |   Name: LFRX (0)
|   |   |   Antennas: AB, BA, A, B
|   |   |   Sensors: 
|   |   |   Freq range: -32.000 to 32.000 MHz
|   |   |   Gain Elements: None
|   |   |   Bandwidth range: 64000000.0 to 64000000.0 step 0.0 Hz
|   |   |   Connection Type: IQ
|   |   |   Uses LO offset: No
|   |     _____________________________________________________
|   |    /
|   |   |       RX Frontend: 1
|   |   |   Name: LFRX (1)
|   |   |   Antennas: AB, BA, A, B
|   |   |   Sensors: 
|   |   |   Freq range: -32.000 to 32.000 MHz
|   |   |   Gain Elements: None
|   |   |   Bandwidth range: 64000000.0 to 64000000.0 step 0.0 Hz
|   |   |   Connection Type: IQ
|   |   |   Uses LO offset: No

**Reply by AlexandreRouma** [reply]
This is strange. I assume something goofy is being passed to SDR++ causing it to try to create multiple entries for either the samplerate or bandwidth. Without having a LFRX board here it's hard to say because I'd need to see exactly what the API returns. If you know your way around C/C++ you could add a line or two in the `selectChannel()` function of the `source_modules/usrp_source/src/main.cpp` file to print out the ranges for samplerate and banwidth returned by libUHD (lines 138 and 159).

**Reply by PripyatAutomations** [reply]
Thanks again for taking a look at this.
threw in some fprintf's
        // Get bandwidth ranges
        bandwidths.clear();
        fprintf(stderr, "bwclear\n");
        bandwidths.define(0, "Auto", 0);
        fprintf(stderr, "bw define auto\n");
        uhd::meta_range_t bwRange = dev->get_rx_bandwidth_range(chanId);
        fprintf(stderr, "get bw range\n");

        for (const auto& r : bwRange) {
            fprintf(stderr, "loop: %d\n", r);
            double step = (r.step() == 0.0) ? 100e3 : r.step();
            for (double i = r.start(); i <= r.stop(); i += step) {
                bandwidths.define((int)i, getBandwdithScaled(i), i);
                fprintf(stderr, "bwidth: %f\n", i);
            }
        }


[23/05/2023 20:00:52.000] [INFO] SoapyModule 'SoapySDR Source': Menu Deselect!
samprate: 8000.000000
samprate: 16000.000000
samprate: 22050.000000
samprate: 32000.000000
samprate: 44100.000000
samprate: 48000.000000
samprate: 96000.000000
samprate: 192000.000000
antennas: OK
gainrange: OK
bwclear
bw define auto
get bw range
loop: 1
terminate called after throwing an instance of 'std::runtime_error'
  what():  Key already exists
Aborted

**Comment by PripyatAutomations** [comment]
moving the fprintf in the inner loop up to before the faulting line,                 bandwidths.define((int)i, getBandwdithScaled(i), i);,  i get bwidth: 0.000000

**Reply by AlexandreRouma** [reply]
move `fprintf(stderr, "bwidth: %f\n", i);` before `bandwidths.define((int)i...` and remove `fprintf(stderr, "loop: %d\n", r);` because `r` is not an integer, it's a class containing a few different values.

**Reply by AlexandreRouma** [reply]
nvm I'm blind, you just did that

**Reply by AlexandreRouma** [reply]
It seems to be a bug with UHD, returning a 0 bandwidth makes no sense at all. I'll add a work around

**Reply by AlexandreRouma** [reply]
Actually, I'd need to know a bit more, could you replace `fprintf(stderr, "bwidth: %f\n", i);` with `fprintf(stderr, "bwidth: %lf -> %lf\n", r.start(), r.stop());`. Should print the exact range.

**Reply by PripyatAutomations** [reply]
I threw an if (> 0) in there. Seems work. Thanks! Saved me many hours

**Reply by AlexandreRouma** [reply]
Well it's a work around in your case but I'd like to fix it properly so that it doesn't happen on other hardware.

**Reply by PripyatAutomations** [reply]
Absolutely. I've wandered outside for some fire with friends but I'll make the requested changes and post output in a few hours. Thanks for the assistance. Saved me a lot of time getting it going!

**Comment by PripyatAutomations** [comment]
A little more poking at it and interestingly after removing the if (>0) it still works *until* i change devices in the devices drop down to any of the strange soundcard devices. then we crash..

Problem seems to have come from the first of those happens to be the default until get in far enough to select the actual USRP device. by adding that if (...) continue in there

this is what i get in the drop down.  -- Any of the soapy ... options will crash sdr++.

![image](https://github.com/AlexandreRouma/SDRPlusPlus/assets/84538339/fbbeeb56-5d8e-4105-9d55-2403b8f2ce4b)

**Reply by AlexandreRouma** [reply]
I'm not sure what those soapy devices are because I don't have them on either linux or windows but they return junk data.

**Reply by PripyatAutomations** [reply]
They're sound card devices in pulse audio. They show up in gqrx, etc too, it just seems to avoid them.
